DRILL-4704 fix

daveoshinsky opened the pull request:
calculate precision from integer value, instead of using faulty
precision of zero

amansinha100 commented:
This patch is missing a unit test..can you add a unit test for this in TestDecimal.java ?

daveoshinsky commented:
I'm getting errors like this one even from unit tests in TestDecimal that I have not modified.  Anyone have an explanation?
Tests run: 7, Failures: 0, Errors: 7, Skipped: 0, Time elapsed: 25.801 sec <<< FAILURE! - in org.apache.drill.exec.physical.impl.TestDecimaltestCastFromFloat(org.apache.drill.exec.physical.impl.TestDecimal)  Time elapsed: 6.876 sec  <<< ERROR!^Morg.apache.drill.exec.rpc.RpcException: org.apache.drill.common.exceptions.UserRemoteException: SYSTEM ERROR: PatternSyntaxException: Unexpected internal error near index 1\ ^

[Error Id: 2b33914f-2262-4370-8d1b-b5e554b02c4c on DaveOshinsky-PC.gp.cv.commvault.com:31010]        at org.apache.drill.exec.rpc.RpcException.mapException(RpcException.java:60)        at org.apache.drill.exec.client.DrillClient$ListHoldingResultsListener.getResults(DrillClient.java:424)        at org.apache.drill.exec.client.DrillClient.runQuery(DrillClient.java:321)        at org.apache.drill.exec.physical.impl.TestDecimal.testCastFromFloat(TestDecimal.java:102)Caused by: org.apache.drill.common.exceptions.UserRemoteException: SYSTEM ERROR: PatternSyntaxException: Unexpected internal error near index 1\ ^
 

    On Friday, June 17, 2016 10:05 AM, Aman Sinha <notifications@github.com> wrote:
 

 This patch is missing a unit test..can you add a unit test for this in TestDecimal.java ?—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

amansinha100 commented:
The tests passed for me (I am running on a linux box).  

mvn install -DskipTests  
mvn -Dtest=TestDecimal surefire:test -DfailIfNoTests=false

Running org.apache.drill.exec.physical.impl.TestDecimal#testCastFromFloat
Running org.apache.drill.exec.physical.impl.TestDecimal#testComplexDecimal
Running org.apache.drill.exec.physical.impl.TestDecimal#testSimpleDecimalArithmetic
Running org.apache.drill.exec.physical.impl.TestDecimal#testComplexDecimalSort
Running org.apache.drill.exec.physical.impl.TestDecimal#testSimpleDecimalMathFunc
Running org.apache.drill.exec.physical.impl.TestDecimal#testSimpleDecimal
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 27.566 sec - in org.apache.drill.exec.physical.impl.TestDecimal

Results :

Tests run: 6, Failures: 0, Errors: 0, Skipped: 0

I see control chars in your error stack .. what platform are you using ? Is it windows ?

daveoshinsky commented:
Yes, Windows 7.  java version "1.8.0_91",  Maven 3.3.9, running build under cygwin
doshinsky@DaveOshinsky-PC /cygdrive/c/apache/drill-1.6.0/exec/java-exec$ mvn -Dtest=TestDecimal surefire:test -DfailIfNoTests=false[INFO] Scanning for projects...[INFO][INFO] ------------------------------------------------------------------------[INFO] Building exec/Java Execution Engine 1.6.0[INFO] ------------------------------------------------------------------------....Running org.apache.drill.exec.physical.impl.TestDecimal#testComplexDecimalRunning org.apache.drill.exec.physical.impl.TestDecimal#testCastFromIntRunning org.apache.drill.exec.physical.impl.TestDecimal#testSimpleDecimalArithmeticRunning org.apache.drill.exec.physical.impl.TestDecimal#testComplexDecimalSortRunning org.apache.drill.exec.physical.impl.TestDecimal#testSimpleDecimalMathFuncRunning org.apache.drill.exec.physical.impl.TestDecimal#testSimpleDecimalTests run: 7, Failures: 0, Errors: 7, Skipped: 0, Time elapsed: 30.062 sec <<< FAILURE! - in org.apache.drill.exec.physical.impl.TestDecimaltestCastFromFloat(org.apache.drill.exec.physical.impl.TestDecimal)  Time elapsed: 8.202 sec  <<< ERROR!org.apache.drill.exec.rpc.RpcException: org.apache.drill.common.exceptions.UserRemoteException: SYSTEM ERROR: PatternSyntaxException: Unexpected internal error near index 1\ ^

[Error Id: 8a0fff12-6dea-407d-99f2-12ea0479b0f7 on DaveOshinsky-PC.gp.cv.commvault.com:31010]        at org.apache.drill.exec.rpc.RpcException.mapException(RpcException.java:60)        at org.apache.drill.exec.client.DrillClient$ListHoldingResultsListener.getResults(DrillClient.java:424)        at org.apache.drill.exec.client.DrillClient.runQuery(DrillClient.java:321)        at org.apache.drill.exec.physical.impl.TestDecimal.testCastFromFloat(TestDecimal.java:111)Caused by: org.apache.drill.common.exceptions.UserRemoteException: SYSTEM ERROR: PatternSyntaxException: Unexpected internal error near index 1\ ^
 

    On Friday, June 17, 2016 2:10 PM, Aman Sinha <notifications@github.com> wrote:
 

 The tests passed for me (I am running on a linux box). mvn install -DskipTests

mvn -Dtest=TestDecimal surefire:test -DfailIfNoTests=falseRunning org.apache.drill.exec.physical.impl.TestDecimal#testCastFromFloat
Running org.apache.drill.exec.physical.impl.TestDecimal#testComplexDecimal
Running org.apache.drill.exec.physical.impl.TestDecimal#testSimpleDecimalArithmetic
Running org.apache.drill.exec.physical.impl.TestDecimal#testComplexDecimalSort
Running org.apache.drill.exec.physical.impl.TestDecimal#testSimpleDecimalMathFunc
Running org.apache.drill.exec.physical.impl.TestDecimal#testSimpleDecimal
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 27.566 sec - in org.apache.drill.exec.physical.impl.TestDecimalResults :Tests run: 6, Failures: 0, Errors: 0, Skipped: 0I see control chars in your error stack .. what platform are you using ? Is it windows ? —
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

amansinha100 commented:
unfortunately, I don't have access to a windows machine.  I would suggest a few options:
1. I just noticed that your test run shows 7 tests.  Mine had 6 tests.  I suppose you added a new one.  Can you first try a clean version without your new test ? 
2. Could you try it on a different machine with either mac or linux ?  
3. Maybe you can seek help from others in the community (I believe @adityakishore runs on windows + cygwin).

daveoshinsky commented:
If I have some extra time to set this up on a Mac or Linux, I will.  That's not so convenient at the moment.  Besides, this should work on Windows, shouldn't it?  I noticed that there's an RpcException coming in from the server side (the drill-bit), but the exception includes nothing like a stack trace to figure out where it's failing.  How can I debug the server side?  Is there a log file?  Or which source file can I change to catch the exception on server side, and log a stack trace?
Regarding the 7th test, yes I added it.  I tried a fresh git clone to Windows from apache/drill master, and that TestDecimal (with 6 tests) fails exactly the same way.  It does appear to have something to do with the different newline convention on Windows (CR-LF), but I need something like a server-side stack trace to localize the problem.  I tried hacking TestDecimal.java to change all \n to \r\n (CR LF) in the query string that is sent to the server (Windows only), but that did not help. 

    On Friday, June 17, 2016 6:01 PM, Aman Sinha <notifications@github.com> wrote:
 

 unfortunately, I don't have access to a windows machine. I would suggest a few options:
1. I just noticed that your test run shows 7 tests. Mine had 6 tests. I suppose you added a new one. Can you first try a clean version without your new test ? 
2. Could you try it on a different machine with either mac or linux ?

3. Maybe you can seek help from others in the community (I believe @adityakishore runs on windows + cygwin). —
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

amansinha100 commented:
Sorry for the delay.  Can you force the following flag to TRUE to get the more detailed stack trace ? 
https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java#L262
(normally, one would do an 'alter session ...'  but TestDecimal is doing a plan submission instead of sql string).

daveoshinsky commented:
In the mean time, I switched over to a Mac, where all 6 existing TestDecimal unit tests worked.  I added a new unit test testCastFromInt patterned after testCastFromFloat, with some code to print the query that it's sending to the DrillBit.  It's failing as shown below.  What did I do wrong?
Query: {  "head" : {    "version" : 1,    "generator" : {      "type" : "org.apache.drill.exec.planner.logical.DrillImplementor",      "info" : ""    },    "type" : "APACHE_DRILL_PHYSICAL",    "resultMode" : "EXEC"  },  graph:[  {      @id:1,      pop:"fs-scan",      format: {type: "json"},      storage:{type: "file", connection: "classpath:///"},      files:["/input_simple_decimal.json"]  }, {    "pop" : "project",    "@id" : 2,    "exprs" : [ {      "ref" : "I4",      "expr" : " (cast(DEC9 as INTEGER)) "    },    { "ref" : "I8", "expr": "(cast(DEC18 as INTEGER))" }    ],
    "child" : 1  },{    "pop" : "project",    "@id" : 4,    "exprs" : [ {      "ref" : "DECIMAL_9",      "expr" : " cast(I4 as decimal9(9, 0))  "    },    {"ref": "DECIMAL38", "expr" : "cast(I8 as decimal38sparse(38, 0))"}    ],
    "child" : 2  },{    "pop" : "screen",    "@id" : 5,    "child" : 4  } ]}
org.apache.drill.exec.rpc.RpcException: org.apache.drill.common.exceptions.UserRemoteException: SYSTEM ERROR: ExpressionParsingException: Expression has syntax error! line 1:15:no viable alternative at input 'INTEGER'

[Error Id: a7d6acc9-70ee-4e6d-9002-90715b8de351 on 10.0.0.211:31010] at org.apache.drill.exec.rpc.RpcException.mapException(RpcException.java:60) at org.apache.drill.exec.client.DrillClient$ListHoldingResultsListener.getResults(DrillClient.java:479) at org.apache.drill.exec.client.DrillClient.runQuery(DrillClient.java:329) at org.apache.drill.exec.physical.impl.TestDecimal.runQuery(TestDecimal.java:58) at org.apache.drill.exec.physical.impl.TestDecimal.testCastFromInt(TestDecimal.java:180)
Here's the new function in TestDecimal.java:
    @Test    public void testCastFromInt() throws Exception {
        // Function checks for casting from Float, Double to Decimal data types        try (RemoteServiceSet serviceSet = RemoteServiceSet.getLocalServiceSet();             Drillbit bit = new Drillbit(CONFIG, serviceSet);             DrillClient client = new DrillClient(CONFIG, serviceSet.getCoordinator())) {
            // run query.            bit.run();            client.connect();            enableDecimalDataType(client);            List<QueryDataBatch> results = runQuery(client, "/decimal/cast_int_decimal.json", "/input_simple_decimal.json");
            RecordBatchLoader batchLoader = new RecordBatchLoader(bit.getContext().getAllocator());
            QueryDataBatch batch = results.get(0);            assertTrue(batchLoader.load(batch.getHeader().getDef(), batch.getData()));

            // the first pair of values in the input consists of integers.  ignore the rest.            String decimal9Output[] = {"99.0000"};            String decimal38Output[] = {"123456789.0000"};
            Iterator<VectorWrapper<?>> itr = batchLoader.iterator();
            // Check the output of decimal9            ValueVector.Accessor dec9Accessor = itr.next().getValueVector().getAccessor();            ValueVector.Accessor dec38Accessor = itr.next().getValueVector().getAccessor();

            for (int i = 0; i < decimal9Output.length; i++) {                assertEquals(dec9Accessor.getObject(i).toString(), decimal9Output[i]);                assertEquals(dec38Accessor.getObject(i).toString(), decimal38Output[i]);            }            assertEquals(6, dec9Accessor.getValueCount());            assertEquals(6, dec38Accessor.getValueCount());
            batchLoader.clear();            for (QueryDataBatch result : results) {              result.release();            }        }    }
    public List<QueryDataBatch> runQuery(DrillClient client, String cast, String input) throws Exception {        String s1 = Files.toString(FileUtils.getResourceAsFile(cast), Charsets.UTF_8);        String s2 = s1.replace("#{TEST_FILE}", input);        if (isWindows()) {            s2 = s2.replace("\n", "\r\n");        }System.out.println("Query: " + s2);  // DAO DEBUG        List<QueryDataBatch> results;        try {            results = client.runQuery(org.apache.drill.exec.proto.UserBitShared.QueryType.PHYSICAL, s2);        }        catch (Exception ex) {            ex.printStackTrace();            Throwable cause = ex.getCause();            if (cause != null) {                System.out.println("Caused by: " + cause.toString());                cause.printStackTrace();             }             throw ex;        }        return results;    }

Here's the accompanying new json file cast_int_decimal.json:
{  "head" : {    "version" : 1,    "generator" : {      "type" : "org.apache.drill.exec.planner.logical.DrillImplementor",      "info" : ""    },    "type" : "APACHE_DRILL_PHYSICAL",    "resultMode" : "EXEC"  },  graph:[  {      @id:1,      pop:"fs-scan",      format: {type: "json"},      storage:{type: "file", connection: "classpath:///"},      files:["#{TEST_FILE}"]  }, {    "pop" : "project",    "@id" : 2,    "exprs" : [ {      "ref" : "I4",      "expr" : " (cast(DEC9 as INTEGER)) "    },    { "ref" : "I8", "expr": "(cast(DEC18 as INTEGER))" }    ],
    "child" : 1  },{    "pop" : "project",    "@id" : 4,    "exprs" : [ {      "ref" : "DECIMAL_9",      "expr" : " cast(I4 as decimal9(9, 0))  "    },    {"ref": "DECIMAL38", "expr" : "cast(I8 as decimal38sparse(38, 0))"}    ],
    "child" : 2  },{    "pop" : "screen",    "@id" : 5,    "child" : 4  } ]}
 

    On Sunday, June 19, 2016 11:08 PM, Aman Sinha <notifications@github.com> wrote:
 

 Sorry for the delay. Can you force the following flag to TRUE to get the more detailed stack trace ? 
https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java#L262
(normally, one would do an 'alter session ...' but TestDecimal is doing a plan submission instead of sql string). —
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

daveoshinsky commented:
I saw no emails yesterday associated with my adding a unit test here for DRILL-4704 fix.  Just in case those got lost somewhere, please review.

amansinha100 commented:
Thanks for adding the tests.  I will take a look likely by tomorrow.

daveoshinsky commented:
Now that 1.7 release dust has hopefully settled down (congrats on release!), please review added unit test.

amansinha100 commented on exec/java-exec/src/main/codegen/templates/Decimal/CastIntDecimal.java:
```diff
@@ -64,7 +64,15 @@ public void setup() {
 
     public void eval() {
         out.scale = (int) scale.value;
-        out.precision = (int) precision.value;
```
I am not too familiar with the decimal type but taking a closer look at the fix I don't think we should ignore the supplied precision.  Your fix will compute the precision in all cases, which would be a performance overhead.  It should be computing only when supplied precision is 0.  Thoughts ?

daveoshinsky replied:
The DRILL-4704 bug root cause was that there is no supplied precision in this case (casting an int to a decimal), i.e., the "supplied precision" was a bogus value of 0, and the casted decimal with precision 0 matched nothing (unless perhaps the value was 0, but I didn't try that).  The most direct way I could fix this was to compute the precision, which just involves a few integer divides by 10 basically.  It might be possible to assume a large enough precision (say 38, to fit most any integer), but that would waste lots of space (to hold a much smaller value than worst case precision of 38 can hold).  Moreover, whenever one assumes a "worst case", eventually it's likely to hit a "worse than worst case" (remember fixed length char buffers in C?  128, no 256, no 1024, and so on).  It's been a while since I ran this (and it was difficult to debug because of the runtime code generation and compilation), but it may be the case that the cast from int to decimal is done only once (to prepare a decimal for comparison, just once), in which case the divisions by 10 really wouldn't make any difference. 

    On Tuesday, July 12, 2016 11:13 AM, Aman Sinha <notifications@github.com> wrote:
 

 In exec/java-exec/src/main/codegen/templates/Decimal/CastIntDecimal.java:> @@ -64,7 +64,15 @@ public void setup() {
>  
>      public void eval() {
>          out.scale = (int) scale.value;
> -        out.precision = (int) precision.value;
I am not too familiar with the decimal type but taking a closer look at the fix I don't think we should ignore the supplied precision. Your fix will compute the precision in all cases, which would be a performance overhead. It should be computing only when supplied precision is 0. Thoughts ? —
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

amansinha100 replied:
The eval() method is called for every row, so there will be a performance overhead.  I  agree about the storage cost if we assume worst case precision; however the per-row evaluation should be as lightweight as possible and likely outweighs the storage consideration.  One option is to treat the IntDecimal and BigIntDecimal separately.  For IntDecimal the max int value is 2^31 - 1 which has a max precision of 10.  Similarly, there is an upper bound on the BigIntDecimal.  You could initialize the precision in the setup() method which is called *once*  per new schema (in the normal case the schema of that column does not change, this would avoid unnecessarily computing the precision).

daveoshinsky replied:
I tried doing it in the setup method, and it did not work. It seems the integer value to be casted was not available at that time.  Since the number of divisions by 10 (these are cheap integer divisions) is proportional to log10(the integer number), I don't see how this is expensive, especially compared to the expense of reading the rows from whatever device. 

    On Tuesday, July 12, 2016 1:56 PM, Aman Sinha <notifications@github.com> wrote:
 

 In exec/java-exec/src/main/codegen/templates/Decimal/CastIntDecimal.java:> @@ -64,7 +64,15 @@ public void setup() {
>  
>      public void eval() {
>          out.scale = (int) scale.value;
> -        out.precision = (int) precision.value;
The eval() method is called for every row, so there will be a performance overhead. I agree about the storage cost if we assume worst case precision; however the per-row evaluation should be as lightweight as possible and likely outweighs the storage consideration. One option is to treat the IntDecimal and BigIntDecimal separately. For IntDecimal the max int value is 2^31 - 1 which has a max precision of 10. Similarly, there is an upper bound on the BigIntDecimal. You could initialize the precision in the setup() method which is called once per new schema (in the normal case the schema of that column does not change, this would avoid unnecessarily computing the precision). —
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

daveoshinsky commented:
Regarding Aman's last comment, please add the following... In the scenario addressed by this fix, there is a const integer that came from the SQL select where clause, and that is what is being casted into a decimal, for comparison with each a decimal value in each row.   Since the integer is constant, a thoroughly optimized piece of code would cast it just once to a decimal, and use the stored decimal to compare with the decimal value in each row.  It appears that that is not what this code is doing. It is casting repeatedly, for each row.  So, rather than be concerned with just a few added integer divisions by 10 (a small number around log10(the const integer) + 1 divides), this code should ideally be (more) extensively modified to do the cast operation just once.  That's too big a change for me to make for this fix.  Besides, which is preferable?  Code that does not function properly (DRILL-4704), and is not well optimized?  Or code that behaves correctly, and is only slightly less well optimized?

daveoshinsky commented:
I added a print to System.out for each time that the cast (including the precision calculation I added) is called.  I found that the cast is done once for each input integer value from the SQL select statement.  It is NOT done once per row in the Parquet file(s) that Drill is scanning.  So, there should be absolutely no performance issue with doing the precision calculation as I am doing it.  Again, this is NOT done for every row.  Please review.

jinfengni commented:
I agreed with Aman that the eval() of CastIntDecimal28Sparse() should not ignore the specified scale/precision, if they are passed as parameters. 

As you analyzed, the problem is the scale/precision is 0 for a int constant value, which is not right. We should fix place where scale/precision is specified wrongly.  Ignoring scale/precision and re-calcualting them in eval() means Drill has to apply that logic per input row.

The reason that you saw eval() is called only once:
Filter employee_id = 170, 
  is   employee_id = castIntDecimal28Sparse(170, scale, precision). // scale and precision is wrongly 0.

Since RHS is a constant, Drill will evaluate only once, even you have multiple rows. However, this is not true, if it's not a constant expression. 

To me, you point to the right cause of the problem. But we need get better place to apply the fix ( not in eval() ) method. 

The place where we get scale/precision as 0 

https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/expr/ExpressionTreeMaterializer.java#L315-L316

daveoshinsky commented:
Thanks Jinfeng for the very clear explanation, and pointer to ExpressionTreeMaterializer.java.  Rather than change ExpressionTreeMaterializer.java, with which I am completely unfamiliar, I added another change to CastIntDecimal.java to only calculate the precision when the supplied one is zero.  I hope the amended fix is sufficient for now.  There might be other cases where the supplied precision is invalid (zero), and this change will handle those. 

    On Wednesday, August 3, 2016 7:18 PM, Jinfeng Ni <notifications@github.com> wrote:
 

 I agreed with Aman that the eval() of CastIntDecimal28Sparse() should not ignore the specified scale/precision, if they are passed as parameters. As you analyzed, the problem is the scale/precision is 0 for a int constant value, which is not right. We should fix place where scale/precision is specified wrongly. Ignoring scale/precision and re-calcualting them in eval() means Drill has to apply that logic per input row.The reason that you saw eval() is called only once:
Filter employee_id = 170, 
 is employee_id = castIntDecimal28Sparse(170, scale, precision). // scale and precision is wrongly 0.Since RHS is a constant, Drill will evaluate only once, even you have multiple rows. However, this is not true, if it's not a constant expression. To me, you point to the right cause of the problem. But we need get better place to apply the fix ( not in eval() ) method. The place where we get scale/precision as 0 https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/expr/ExpressionTreeMaterializer.java#L315-L316—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

jinfengni commented:
> I attached debugger to Drill and found it's passing through line 172 (not 315-316) of ExpressionTreeMaterializer.java, and that the "toType" object is one of those runtime generated ones (name like TypeProtos$MajorType@8570).  Rather than attempt to change that code, I made another change to CastIntDecimal.java which avoids repeated calculation of the precision when it is already provided.  Please review.

Line 172 was called after scale/precision are set on line 315-316.

I understand your new change will kick in when the passed precision= 0. However, I still think that the right way is to pass in the correct scale/precision when call that cast function, in stead of figuring out on-the-fly in the middle of that function.  If someone pass a 0 precision (which is not right), we need fix that in the place where 0 precision is specified.

Also, will your code work correct for negative integer input?

daveoshinsky commented:
No, negative precision will not work properly.  I committed another change to PR-517 to handle this.
Quite honestly, I do not know how to fix the problem as you suggest in ExpressionTreeMaterializer.java.  First, I set a breakpoint on line 315, and it never fired.  It went to the breakpoint on line 172, never passing line 315 (where I set a breakpoint).  I see that code at line 172 is making a call to a virtual function toType.getPrecision() to obtain the precision, from an object of a type that is only known at runtime.  Based on what I saw in the debugger, the type of the "toType" object at line 172 was something like TypeProtos$MajorType@8570, which looked like a piece of runtime generated code (as often seen in Drill), and the debugger would not let me single step into it.  That toType object did have both scale and precision = 0.  Now, if I could obtain the integer value of the toType object, I could calculate the precision right there (if the getPrecision() returns 0).  But, I don't see how to get the integer value at line 172.  In fact, I'm guessing that "getting an integer value" only applies in some cases, for some kinds of values that are being "materialized".
To summarize, I have not been looking at Apache Drill nearly as long as you have, and don't understand it nearly as well as you.  Yet, you are asking me to change a fix that already works, to fit a model you appear to have in mind for a fix as you would do it.  I simply don't have the level of understanding of the Drill code to do the fix as you imagine it should be done.  Moreover, I am a volunteer, being paid by my employer to work mainly on things other than Apache Drill.  As such, the amount of additional time I can spend on this fix is limited.  My fix already works AFAIK, and is not inefficient.  I suggest you take it, and if necessary, then change it to your liking.  The unit test I've included is a decent test of the DRILL-4704 problem. 

    On Thursday, August 4, 2016 4:06 PM, Jinfeng Ni <notifications@github.com> wrote:
 

 
I attached debugger to Drill and found it's passing through line 172 (not 315-316) of ExpressionTreeMaterializer.java, and that the "toType" object is one of those runtime generated ones (name like TypeProtos$MajorType@8570). Rather than attempt to change that code, I made another change to CastIntDecimal.java which avoids repeated calculation of the precision when it is already provided. Please review.
Line 172 was called after scale/precision are set on line 315-316.I understand your new change will kick in when the passed precision= 0. However, I still think that the right way is to pass in the correct scale/precision when call that cast function, in stead of figuring out on-the-fly in the middle of that function. If someone pass a 0 precision (which is not right), we need fix that in the place where 0 precision is specified.Also, will your code work correct for negative integer input?—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

paul-rogers commented:
The proximate cause of this particular issue is the mechanism by which the cast function is being generated. As others have noted, the code is passing through lines 316-316. My guess is that the materializer is working out how to call this function:

Decimal28SparseFunctions.Decimal28SparseEqual( Decimal28Sparse left, Decimal28Sparse right )
The materializer sees: D28SEqual( Decimal28Sparse, Integer )

The code wants to convert the Integer to a Decimal28Sparse. It asks the argument for its precision and scale. But, since the argument just specifies the "minor type" (SQL type), it does not have the extended attributes of precision and scale. Thus, the annotations in the function definition are the ultimate cause of the precision being 0:

@Param Decimal28SparseHolder left;
@Param Decimal28SparseHolder right;

Now, we might argue that the code should be using, as the target (p, s) the (p,s) of the left argument (the one that is already a decimal). But, the materializer does not know that this is an operator; it treats all functions the same (operators and otherwise.) So, the materializer does not know that we should treat arguments in this way. All it knows is that it has an Integer and wants a D28S (of, presumably, any p & s).

Next, we can ask how DS28Equal handles the (p,s) of the two arguments: must they be the same or can they be different? Looking at the code (as best as I can tell), the code will align values with different (p,s) values. Thus, say, values of Decimal28( 10, 2 ) and Decimal28( 8, 0 ) will compare properly.

So, back to the cast from Integer to D28S. As Dave suggests, we can interpret (precision == 0) to mean "you figure it out": we can set (p,s) way we want.

Dave's fix computes the precision based on the actual integer value. But, since the comparison can handle any valid (p,s), we might as well set them to a constant. Since the maximum Integer value is 2 billion, we can just use a constant (10,0).

If the selected precision is constant, then the cost is low and the solution is fine for both for constant integers (as in the original case), and for per-row conversions (as in Aman's concern.)

paul-rogers commented:
The plot thickens. I tried the fix of setting the precision to a constant of 10. This uncovered a larger issue. The template in question generates cast functions for (INTEGER, BIGINT) x (DECIMAL9, DECIMAL18, DECIMAL28) and perhaps others. The constant of 10 does not, of course, work for BIGINT (long) values.

The trick is that precision=10 won't work for DECIMAL9 either. Dave's solution has a similar problem. Dave sets the precision to whatever is right for the input value, which seems great. But, that value could be too large for the output DECIMAL type.

What we need is to set the precision to the min( max int precision, max decimal precision ). Or, if we use Dave's proposed solution, max( input arg precision, max decimal precision ).

In either case, the code must handle overflow. Passing a Long.MAX_VALUE or even Integer.MAX_VALUE to CastBigIntDecimal9( ) should cause an overflow error or data truncation. I'll research how that worked previously to see if we've uncovered a new issue, or if a solution already exists.

daveoshinsky commented:
Thanks for looking at this, Paul.  If there are problems with comparing an input decimal (from the "where clause", for example) against the data, it is because of the algorithms for comparing decimal values of different precisions, not because the proper precision was computed for the input decimal (e.g., based on the actual numerical value, as my PR-517 fix is doing when no precision is provided initially).  I was going to comment about the idea of using a fixed value of precision=10, but you beat me to it.  It does not suffice for larger integers, like those that require a "long" instead of an "int" to represent.  So, I suggest that we take a "divide and conquer" approach with this problem.  First, compute a proper precision for the value casted from int or BigInt.  Next, make sure the comparison with other decimal values works for any other decimal value precision (a separate issue).
Decimal support in Drill is disabled by default for good reasons.  The (decimal) design itself is fundamentally flawed, and it has all sorts of issues related to code maintainability as well as runtime performance.  I have been experimenting in a Drill "play build node" with having a single, one-size-fits-all variable width vector class that represents any decimal value efficiently (I call it "VarDecimal").  It's difficult to even get this to compile (with all of the FreeMarker use, and the numerous conversions between all of the decimal and non-decimal types), but eventually I hope to experiment with this to deal with the DRILL-4184 problem I encountered a number of months ago (for which I have a pull request PR-372 with a short-term, but not a clean, fix).
As food for thought, while looking at this experimental "VarDecimal" thing, I noticed that there are (numerous) casts/conversions between all of the pairwise combinations of different decimal precisions (DECIMAL9, DECIMAL18, DECIMAL28, DECIMAL38, both dense and sparse, from what I recall), and between all of these and all other types that have numeric interpretation.  Eventually replacing all of these different DECIMAL* fixed width representations with a single VarDecimal variable width representation (one-size-fits-all) would be much more efficient (memory-wise) at runtime (for typical scenarios where most actual numeric values don't require the full precision to represent), would greatly simplify parts of the Drill code, and would fix DRILL-4184 cleanly. 
Until some major re-work can be done for Drill decimal support, we will probably have to settle for small, incremental improvements.  I would say that PR-517 represents one such improvement.

    On Thursday, August 4, 2016 9:31 PM, Paul Rogers <notifications@github.com> wrote:
 

 The plot thickens. I tried the fix of setting the precision to a constant of 10. This uncovered a larger issue. The template in question generates cast functions for (INTEGER, BIGINT) x (DECIMAL9, DECIMAL18, DECIMAL28) and perhaps others. The constant of 10 does not, of course, work for BIGINT (long) values.The trick is that precision=10 won't work for DECIMAL9 either. Dave's solution has a similar problem. Dave sets the precision to whatever is right for the input value, which seems great. But, that value could be too large for the output DECIMAL type.What we need is to set the precision to the min( max int precision, max decimal precision ). Or, if we use Dave's proposed solution, max( input arg precision, max decimal precision ).In either case, the code must handle overflow. Passing a Long.MAX_VALUE or even Integer.MAX_VALUE to CastBigIntDecimal9( ) should cause an overflow error or data truncation. I'll research how that worked previously to see if we've uncovered a new issue, or if a solution already exists.—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

paul-rogers commented:
Hi Dave,

I agree that Drill Decimal support clearly needs attention. I'm expermenting with adjusting the template code to at least generate a meaningful conversion for all int x decimal types when the values fit.

Looks like current behavior will be that if we convert an int that is too large for a scale of 0, we won't adjust the scale. That is, if we always leave scale=0, then we'll do the following obviously wrong converstion:

big int: 123,456,789,012
Dec 9: 456,789,012

We want a scale of 3, so that the answer is:

Dec 9: 123,456,789,000

And, of course, we should round the least significant digit, which it seems we don't do.

So, our fix will work only for the simple case, not for the more complex case.

We could leverage your fix to properly set the scale as well as the precision.

paul-rogers commented:
Based on above discussion, I played around with the proposed solution. The following takes into consideration the max precision of both the input and output types. It DOES NOT handle truncation properly (but nor do any of the other solutions.)

Here is the revised code:

        if (precision.value <= 0) {
    <#--
            // since input precision is nonpositive, calculate precision for this integer value
            long precisionTmp = in.value == 0 ? 1 : in.value;  // precision for value 0 will be 1
            int precisionCounter = 0;
            while (precisionTmp != 0) {
                ++precisionCounter;
                precisionTmp /= 10;
            }
            out.precision = precisionCounter;
    -->
            // Since precision is not provided, select one appropriate for the
            // largest integer value and the target decimal type.

            // Note: if the integer is too large for the target precision, silent
            // truncation of the highest part of the number will occur. This is
            // obviously bad. What should happen is we shift scale and discard
            // the low-order digits.
            // That is, if we convert 123,456,789,012 to Dec9, the result is
            // 456,789,012 which is very obviously very wrong. It should be
            // 123,456,789,000.
            <#-- A correct solution requires a modification of the code
                 commented out above. -->

          <#-- Precision needed by the largest from type value. -->
          <#if type.from.equals( "Int" )>
            <#assign inPrec = 10>
          <#elseif type.from.equals( "BigInt" )>
            <#assign inPrec = 19>
          <#else>
            <#-- Yes, create invalid syntax: that way the compiler will tell us we missed a type. -->
            <#assign inPrec = "Unexpected from type: ${type.from}">
          </#if>

          <#-- Maximum precision allowed by the to type. -->
          <#if type.to.startsWith("Decimal9")>
            <#assign maxPrec = 9>
          <#elseif type.to.startsWith("Decimal18")>
            <#assign maxPrec = 18>
          <#elseif type.to.startsWith("Decimal28")>
            <#assign maxPrec = 28>
          <#elseif type.to.startsWith("Decimal38")>
            <#assign maxPrec = 38>
          <#else>
            <#-- Yes, create invalid syntax: that way the compiler will tell us we missed a type. -->
            <#assign maxPrec = "Unexpected to type: ${type.to}">
          </#if>
          <#-- Note that this calculation is done here, rather than in static variables,
               because static members are not allowed for this function. -->
          <#if inPrec < maxPrec>
            // Maximum precision needed by the largest ${type.from} value.
            out.precision = ${inPrec};
          <#else>
            // Maximum precision allowed by the ${type.to} type.
            out.precision = ${maxPrec};
          </#if>
        }
        else {
            // since input precision is positive, assume it is correct, and use it
            out.precision = (int) precision.value;
            out.scale = (int) scale.value;
        }

paul-rogers commented:
As a next step, using Dave's solution would better handle truncation: based on the precision of the input value, adjust the scale to truncate low-order digits.

paul-rogers commented:
To clarify the previous comment... In the case where the maximum required precision for the int type is greater than the maximum precision of the target decimal type, we use Dave's code to figure out the actual precision of the int value. Then, we figure out scale as:

    scale = value precision - max decimal precision

This, if you try to store 123,456,789,012 into an Dec9, you would get 123,456,789,000 (truncate the lower-order digits.) The solution does not handle rounding, but that can be added.

Note that the decision to use Dave's code would be in the template: it would apply for, say, BigInt to Dec9 and that kind of thing. The logic to detect this case is already in the suggested code snippet.

      <#if inPrec < maxPrec>
        ...
      <#else>
        // Put Dave's find-the-precision logic here
        if ( out.precision > ${maxPrec} ) {
            out.scale = out.precision - ${maxPrec};
            out.precision = ${maxPrec};
        }
      </#endif>

I have not tested the above; it is just a pseudocode suggestion.

daveoshinsky commented:
The scale is a schema-level property of that column of data, that is never changed on the fly, to deal with problematic truncation, or otherwise.  If anybody is interested in a detail explanation, or other tidbits about how decimal types work, just let me know (daveoshinsky@yahoo.com). 

    On Saturday, August 6, 2016 3:27 PM, Paul Rogers <notifications@github.com> wrote:
 

 To clarify the previous comment... In the case where the maximum required precision for the int type is greater than the maximum precision of the target decimal type, we use Dave's code to figure out the actual precision of the int value. Then, we figure out scale as:scale = value precision - max decimal precision
This, if you try to store 123,456,789,012 into an Dec9, you would get 123,456,789,000 (truncate the lower-order digits.) The solution does not handle rounding, but that can be added.Note that the decision to use Dave's code would be in the template: it would apply for, say, BigInt to Dec9 and that kind of thing. The logic to detect this case is already in the suggested code snippet.  <#if inPrec < maxPrec>
    ...
  <#else>
    // Put Dave's find-the-precision logic here
    if ( out.precision > ${maxPrec} ) {
        out.scale = out.precision - ${maxPrec};
        out.precision = ${maxPrec};
    }
  </#endif>
I have not tested the above; it is just a pseudocode suggestion. —
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

paul-rogers commented:
Good point. So, we'd want to issue an overflow error in the cast method, but it seems Drill has no way to do that. So, for now, maybe just leave it at the version that works when numbers are in range, and file a JIRA for the out-of-range case.

paul-rogers commented:
As a follow-up, the case we're focusing on is an implicit cast: one created by the system to convert an integer to a decimal type in order to do an equality check. To avoid the scale issue, the code that does the explcit cast should check the type of integer and select a target decimal type that will hold all values of that int type with a scale of 0. Does the code do that now? It may be that the decimal type must also be cast so that the equality has the same types. That is, if we compare Dec9 with Integer, we need a scale of 10 for the Int, so we must cast the Int to Dec18. As a result, we must cast the Dec9 to Dec18.

The reason for all of this is that the casts are done statically. A more efficient solution would be for the the run-time code to make a decision based on actual values, but Drill is not set up that way. Another solution would be to have an in-memory form of a decimal that caries the precision and scale so we can set the scale per value. But, Drill is not set up that way either.

So, the change we've been discussing (setting scale based on int and decimal types, ignoring overflow for now) seems the only short-term solution for 1.8.

jinfengni commented:
The change to the cast function implementation will impact not only implicit cast, but also explicit cast. You are right that the test case of employee_id = 170 will trigger implicit cast. However, if you have an explicit cast  "select cast(int_column as decimal)  from ...", you will also see the impact of the code change.  

I could not understand what you mean by "casts are done statically".  For implicit cast, the cast function is inserted dynamically in execution time, only after Drill knows the incoming batch schema.

paul-rogers commented:
"Dynamic" meant "for each data value." "Static" meant "at compile time." We make our type decisions at compile time. That is too early to adjust scale based on actual input values. It is a real nuisance that scale is associated with the type, not the value.

Dave filed a bug to track the additional issues. See DRILL-4834.

paul-rogers commented:
What is the resolution here? Dave, do you want to review & adopt the suggested code several posts back? That code passed the unit tests that you provided, plus those aready in Drill (using the unit tests you suggested in an earlier post.) Then, update this PR with the results. Sounds like Jinfeng can then review the code & accept the PR. As we've said, the fix is not perfect, but it does work for every int/decimal type if the user is careful to provide int values that fit within the target decimal type.

daveoshinsky commented:
Paul, do you mean the long piece of code with conditions on various destination precision types, that sets inPrec and maxPrec, and then selects the smaller of the two for the resulting precision?  I think it's just way simpler to use the calculated precision, as I suggested earlier.  That also works, and it's less code to look at, and less code to potentially be broken.  I do have a bias against the heavy, heavy use of FreeMarker everywhere.  It's just a recipe for bugs piled on top of bugs, and all of the numerous FreeMarker conditions are not likely to be tested for quite some time. 

    On Monday, August 8, 2016 1:48 PM, Paul Rogers <notifications@github.com> wrote:
 

 What is the resolution here? Dave, do you want to review & adopt the suggested code several posts back? That code passed the unit tests that you provided, plus those aready in Drill (using the unit tests you suggested in an earlier post.) Then, update this PR with the results. Sounds like Jinfeng can then review the code & accept the PR. As we've said, the fix is not perfect, but it does work for every int/decimal type if the user is careful to provide int values that fit within the target decimal type.—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

jinfengni commented:
I'm not fully convinced that we should check the precision for each input value for this castIntDecimal function. 

The argument of proposed patch is parameter "precision=0" is not valid, and has to be calculated on-the-fly or "dynamically" for each input value.  To me, 1) if "precision=0" is a wrong input to this function, then we should fix the place where precision=0 is passed in.  2) why would you treat "precision=0" only?  what if I pass in "precision = 1", and a integer value of 123456? in such case, do we think "precison=1" is valid or not? 

Regarding the overflow problem, that seems to be a separate issue, and probably is true for most of the existing decimal functions.

daveoshinsky commented:
The overflow problem will require separate changes to fix, from any changes we make in the short-term to fix this issue (see DRILL-4834).  However, if I understand some of Paul's findings earlier, the two issues are related as follows.  If we choose a large precision based only on the kind of integer that is input to the cast (either an int, or a long, but not minimized to represent the integer value to be casted), then we are more likely to encounter the overflow issue, because this precision is more likely to exceed the capacity of the destination decimal of the cast.  Is that right, Paul?  Given the risk of overflow, it is safer to choose the absolute minimum precision possible, which is what the changes I made to CastIntDecimal.java are doing.  We can't avoid overflow in every situation, but we can try to minimize the chances of it.  My understanding is that the actual integer value to be casted isn't available in ExpressionTreeMaterializer.java; at least, I did not see how to get it.  And that integer value is required to compute the minimum precision to represent that value.  I don't see how to fix this problem reliably in ExpressionTreeMaterializer.java, as I said earlier.  If you (Jinfeng) know how to do that, why don't you just take the changes in PR-517, and then check changes in on top of those to fix the problem your way?  At least, you get my unit test, which was for a condition (DRILL-4704) that was never tested by any earlier unit test. 

    On Monday, August 8, 2016 3:25 PM, Jinfeng Ni <notifications@github.com> wrote:
 

 I'm not fully convinced that we should check the precision for each input value for this castIntDecimal function. The argument of proposed patch is parameter "precision=0" is not valid, and has to be calculated on-the-fly or "dynamically" for each input value. To me, 1) if "precision=0" is a wrong input to this function, then we should fix the place where precision=0 is passed in. 2) why would you treat "precision=0" only? what if I pass in "precision = 1", and a integer value of 123456? in such case, do we think "precison=1" is valid or not? Regarding the overflow problem, that seems to be a separate issue, and probably is true for most of the existing decimal functions. —
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

jinfengni commented:
This is the prototype patch I tried [1]. I replied with the link in our prior direct communication. 

[1] https://github.com/jinfengni/incubator-drill/commits/review/DRILL-4704

daveoshinsky commented:
Got it.  I added a commit, but cleaned up your changes to centralize the duplicated code in a function.  As discussed, this method might increase the chances of an overflow, but it avoids changing the cast.  Also, if there are any other places in the code that pass an invalid (zero) precision to the cast, those will now exhibit the same bug, as the cast is vulnerable (without my earlier fix) to such a bad argument. 

    On Monday, August 8, 2016 3:58 PM, Jinfeng Ni <notifications@github.com> wrote:
 

 This is the prototype patch I tried [1]. I replied with the link in our prior direct communication. [1] https://github.com/jinfengni/incubator-drill/commits/review/DRILL-4704—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

jinfengni commented:
Since we will have a discussion for long term plan of decimal type support (tomorrow hangout), I think it makes sense to apply a small-scoped change to fix this incorrect query result issue. For the overflow problem or decimal type design in general, we could  continue our discussion either in hangout or dev list.

Based on that, I'm inclined to merge this patch. @amansinha100 , could you please take a look at the new change, and see if it addressed your prior concern?

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/expr/ExpressionTreeMaterializer.java:
```diff
@@ -267,6 +267,17 @@ public LogicalExpression visitBooleanOperator(BooleanOperator op, FunctionLookup
+    static public int computePrecision(LogicalExpression currentArg) {
+        int precision = currentArg.getMajorType().getPrecision();
+        if (currentArg.getMajorType().getMinorType() == MinorType.INT) {
+            precision = 10;
```
These values should preferably be defined as static final constants.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/expr/ExpressionTreeMaterializer.java:
```diff
@@ -267,6 +267,17 @@ public LogicalExpression visitBooleanOperator(BooleanOperator op, FunctionLookup
     }
 
+    static public int computePrecision(LogicalExpression currentArg) {
+        int precision = currentArg.getMajorType().getPrecision();
```
Did you consider checking whether the type has a precision or not using getMajorType().hasPrecision()  ?  That way, you would only call getPrecision() if it returned True and otherwise set the precision for INT, BIGINT.

daveoshinsky replied:
Well, if the type has no precision, what do you suggest?  Move the fix back to CastIntDecimal.java (calculate the precision based on the value), as I originally had it, but Jinfeng insisted was incorrect?
You two fellows, Jinfeng and Aman, can decide among yourselves.  Let me know when you're finished duking it out.  If there are any survivors, we can discuss on the hangout tomorrow. 

    On Monday, August 8, 2016 8:04 PM, Aman Sinha <notifications@github.com> wrote:
 

 In exec/java-exec/src/main/java/org/apache/drill/exec/expr/ExpressionTreeMaterializer.java:> @@ -267,6 +267,17 @@ public LogicalExpression visitBooleanOperator(BooleanOperator op, FunctionLookup
>        return new BooleanOperator(op.getName(), args, op.getPosition());
>      }
>  
> +    static public int computePrecision(LogicalExpression currentArg) {
> +        int precision = currentArg.getMajorType().getPrecision();
Did you consider checking whether the type has a precision or not using getMajorType().hasPrecision() ? That way, you would only call getPrecision() if it returned True and otherwise set the precision for INT, BIGINT. —
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

amansinha100 commented:
couple of minor comments; otherwise LGTM.

jinfengni commented:
I'll make minor change to the commit based on Aman's comment, and re-run regression and merge if everything is fine.
